### PR TITLE
[AutoParallel] Recompute Pass

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

New features

### PR changes

Others

### Describe

* add AutoParallel recompute pass.
* this recompute pass modify the complete program (including forward, backward and update), which is different with fleet.meta_optimizers.recompute_optimizer
* the performance is same as fleet.meta_optimizers.recompute_optimizer

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/distributed_passes/test_auto_parallel_recompute_pass.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import paddle.distributed.auto_parallel as auto
+from paddle.distributed.passes import new_pass, PassManager
+from auto_parallel_pass_test_base import AutoPallelPassTestBase
+
+
+class TestRecomputePass(AutoPallelPassTestBase):
+    def init(self):
+        if paddle.is_compiled_with_cuda():
+            paddle.set_flags({'FLAGS_cudnn_deterministic': 1})
+        self.rtol = 1e-5
```
Comment (JZ-LIANG) [inline python/paddle/fluid/tests/unittests/distributed_passes/test_auto_parallel_recompute_pass.py]: 重计算能做到精度逐位对齐吧？

File: python/paddle/fluid/tests/unittests/distributed_passes/auto_parallel_pass_test_base.py [hunk 2]
```diff
@@ -178,13 +172,12 @@ def get_gpt_model(self, strategy, place, batch_size, sequence_len,
...
         criterion = GPTPretrainingCriterion()
         loss = criterion(preds, labels, loss_mask)
-        clip = paddle.nn.ClipGradByNorm(clip_norm=1.0)
         optimizer = paddle.fluid.optimizer.AdamOptimizer(
             learning_rate=0.00001,
             beta1=0.9,
             beta2=0.999,
             epsilon=1e-08,
-            grad_clip=clip)
+            grad_clip=None)
```
Comment (JZ-LIANG) [inline python/paddle/fluid/tests/unittests/distributed_passes/auto_parallel_pass_test_base.py]: 为什么不支持 clip

**Review by JZ-LIANG — DISMISSED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/distributed_passes/test_auto_parallel_recompute_pass.py (hunk 1 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/tests/unittests/distributed_passes/test_auto_parallel_recompute_pass.py]: 基本可以做到逐位对齐，但会偶尔有几个step第6位开始对不齐，误差在1e-16。

File: python/paddle/fluid/tests/unittests/distributed_passes/auto_parallel_pass_test_base.py (hunk 2 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/tests/unittests/distributed_passes/auto_parallel_pass_test_base.py]: 已支持。

File: python/paddle/fluid/tests/unittests/distributed_passes/test_auto_parallel_recompute_pass.py (hunk 1 above)
Comment (zhaoyinglia) [reply python/paddle/fluid/tests/unittests/distributed_passes/test_auto_parallel_recompute_pass.py]: 基本可以做到逐位对齐，但会偶尔有几个step第6位开始对不齐，误差在1e-6

File: python/paddle/distributed/auto_parallel/parallelizer.py [hunk 3]
```diff
@@ -190,7 +193,7 @@ def _get_dist_program(self, rank, dist_context=None, relaunch_phase=False):
         # serial forward pass
         self._apply_pre_optimization_passed(completed_main_program,
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/parallelizer.py]: Rename `_apply_pre_optimization_passed `to `_apply_pre_optimization_passes` and `_apply_post_optimization_passed` to `_apply_post_optimization_passes`.

File: python/paddle/distributed/auto_parallel/reshard.py [hunk 4]
```diff
@@ -26,6 +26,9 @@
 from .dist_attribute import OperatorDistributedAttribute, TensorDistributedAttribute
 from .process_group import new_process_group, ProcessGroup, _g_process_group_map
 
+# NOTE: If op in SPECIAL_OPS, it will not be resharded. 
+SPECIAL_OPS = ['check_finite_and_unscale', 'update_loss_scaling']
```
Comment (aoyulong) [inline python/paddle/distributed/auto_parallel/reshard.py]: The global variable should use `_g_xxxx.` Please rename `SPECIAL_OPS `to `_g_special_ops`.

File: python/paddle/distributed/passes/auto_parallel_recompute.py [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+
+class RecomputeState(ProgramStats):
+    def __init__(self, block, ops):
+        super(RecomputeState, self).__init__(block=block, ops=ops)
+        self._block = block
+        self._ops = ops
+        self.var_op_deps = {}
+
+    def build_stats(self):
```
Comment (aoyulong) [inline python/paddle/distributed/passes/auto_parallel_recompute.py:37]: `build_stats` is `build_state`?

File: python/paddle/distributed/passes/auto_parallel_recompute.py [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+            logging.info("segment start op: [{}]: [{}] [{}]".format(self._ops[
+                idx1].desc.type(), self._ops[idx1].desc.input_arg_names(
+                ), self._ops[idx1].desc.output_arg_names()))
+            logging.info("segment end op: [{}]: [{}] [{}]".format(self._ops[
+                idx2 - 1].desc.type(), self._ops[idx2 - 1].desc.input_arg_names(
+                ), self._ops[idx2 - 1].desc.output_arg_names()))
+
+        return segments
+
+    def modify_forward_desc_for_recompute(self, dist_context):
```
Comment (aoyulong) [inline python/paddle/distributed/passes/auto_parallel_recompute.py:92]: Please add more comments. For example, what is the purpose of `modify_forward_desc_for_recompute`?

**Review by aoyulong — DISMISSED** [review]

File: python/paddle/distributed/passes/auto_parallel_recompute.py (hunk 6 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/passes/auto_parallel_recompute.py:92]: done.

File: python/paddle/distributed/passes/auto_parallel_recompute.py (hunk 5 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/passes/auto_parallel_recompute.py:37]: This function inherits from `ProgramStats` of `backward.py`.

File: python/paddle/distributed/auto_parallel/reshard.py (hunk 4 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/reshard.py]: done.

File: python/paddle/distributed/auto_parallel/parallelizer.py (hunk 3 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/parallelizer.py]: done.

Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/parallelizer.py]: done.

**Review by sneaxiy — APPROVED** [review]
LGTM.

**Pull request merged**